# Linear Programming Optimization Problems (2 Variables)

**Disclaimer:** *This page is a series of questions that look at very basic linear programming optimization problems (with just two variables). The questions have been sourced online, with the links attached in each question. The purpose of this document is for my own, practicing benefits.*

**Before Questions:** Import the gurobipy library

In [38]:
import numpy as np
import gurobipy as gp

## Manufacturing Pants and Jackets Example

Question Source: https://www.superprof.co.uk/resources/academic/maths/linear-algebra/linear-programming/linear-programming-examples.html

A store has requested a manufacturer to produce pants and sports jackets.

For materials, the manufacturer has $750 m ^2$ of cotton textile and $1000 m^2$ of polyester. Every pair of pants (1 unit) needs $1 m^2$ of cotton and $2 m ^2$ of polyester. Every jacket needs $1.5 m ^2$ of cotton and $1 m^2$ of polyester. The price of the pants is fixed at 50 dollars and the jacket, 40 dollars. What is the number of pants and jackets that the manufacturer must give to the stores so that these items obtain a maximum sale?

**Final Equations to Use:** <br>
$ 1X + 1.5Y </= 750 $ Equation 1 <br>
$ 2X + 1Y </= 1000 $ - Equation 2 <br>
$ 50X + 40Y = Profit $ - Equation 3

In [13]:
#create an empty gurobi model
#we will then tell it what the decision variables are along with objective and constraints.
manufacturingMod=gp.Model()

In [4]:
#we first define the variables
#the variables will be used in the objective and constraints
#we have 2 variables here, one for jackets, one for pants
manufacturingModX = manufacturingMod.addMVar(2) 

In [7]:
#using the decision variables, input the objective (equation 3)
manufacturingMod.setObjective(50*manufacturingModX[0] + 40*manufacturingModX[1],sense=gp.GRB.MAXIMIZE)
#by default gurobi wants to minimize.  so we must tell it that we're maximizing

In [8]:
# using the decision variables define the constraints
# first create an empty list that has as many entries as there are constraints
conlist=[0]*2
# then come back and change the list entries one by one to represent each constraint
conlist[0] = manufacturingMod.addConstr(1*manufacturingModX[0] + 1.5*manufacturingModX[1] <= 750)
conlist[1] = manufacturingMod.addConstr(2*manufacturingModX[0] + 1*manufacturingModX[1] <= 1000)

In [10]:
manufacturingMod.Params.OutputFlag = 0 # tell gurobi to shut up!!

manufacturingMod.optimize() # solve the model

In [11]:
manufacturingMod.objVal # print the optimal objective value

28750.0

In [12]:
manufacturingModX.x # print the optimal decision variables

array([375., 250.])

**Answer:**
The maximum profit we can gain is $28,750, by producing 375 pants and 250 jackets.

## Maria's Online Shop Example

Question Source: https://www.superprof.co.uk/resources/academic/maths/linear-algebra/linear-programming/linear-programming-examples.html

Maria has an online shop where she sells hand made paintings and cards. She sells the painting for 50 dollars and the card for $20. It takes her 2 hours to complete 1 painting and 45 minutes to make a single card. She also has a day job and makes paintings and cards in her free time. She cannot spend more than 15 hours a week to make paintings and cards. Additionally, she should make not more than 10 paintings and cards per week.

She makes a profit of 25 dollars on painting and 15 dollars on each card. How many paintings and cards should she make each week to maximize her profit.

**Final Equations to Use:** <br>
$ 120X + 45Y </= 15(60) = 900 $ Equation 1 <br>
$ X + Y </= 10 $ - Equation 2 <br>
$ 50X + 20Y = Profit $ - Equation 3

<br>
Where X = paintings and Y = cards

In [18]:
mariaMod=gp.Model()

In [21]:
mariaModX = mariaMod.addMVar(2)  #paintings and cards

In [22]:
#Equation 3
mariaMod.setObjective(50*mariaModX[0] + 20*mariaModX[1],sense=gp.GRB.MAXIMIZE)

In [33]:
conlist=[0]*2
#Equation 1
conlist[0] = mariaMod.addConstr(120*mariaModX[0] + 45*mariaModX[1] <= 900)
#Equation 2
conlist[1] = mariaMod.addConstr(mariaModX[0] + mariaModX[1] <= 10)


In [34]:
mariaMod.Params.OutputFlag = 0
mariaMod.optimize()

In [35]:
mariaMod.objVal

380.0

In [36]:
mariaModX.x 

array([6., 4.])

**Answer:**
The maximum profit Maria can gain is $380, by producing 6 paintings and 4 cards, per week.

## Calculator Company Example

Question Source: https://www.purplemath.com/modules/linprog3.htmA 

A calculator company produces a scientific calculator and a graphing calculator. Long-term projections indicate an expected demand of at least 100 scientific and 80 graphing calculators each day. Because of limitations on production capacity, no more than 200 scientific and 170 graphing calculators can be made daily. To satisfy a shipping contract, a total of at least 200 calculators much be shipped each day.
 
If each scientific calculator sold results in a 2 dollar loss, but each graphing calculator produces a $5 profit, how many of each type should be made daily to maximize net profits?

**Final Equations to Use:** <br>
$ 100 </= X </= 200 $ Equation 1 <br>
$ 80 </= Y </= 170 $ - Equation 2 <br>
$ X + Y >/= 200 $ - Equation 3 <br>
Profit = 5Y - 2X (Maximize)


<br>
Where X = scientific and Y = graphing

In [37]:
calcMod=gp.Model()

In [39]:
calcModX = calcMod.addMVar(2)  #scientific and graphing

In [47]:
calcMod.setObjective(5*calcModX[1] - 2*calcModX[0],sense=gp.GRB.MAXIMIZE) #profit equation

In [48]:
conlist=[0]*5 #for five equations
#Equation 1
conlist[0] = calcMod.addConstr(calcModX[0] <= 200)
conlist[1] = calcMod.addConstr(calcModX[0] >= 100)

#Equation 2
conlist[2] = calcMod.addConstr(calcModX[1] <= 170)
conlist[3] = calcMod.addConstr(calcModX[1] >= 80)

#Equation 3
conlist[4] = calcMod.addConstr(calcModX[0] + calcModX[1] >= 200)


In [49]:
calcMod.Params.OutputFlag = 0
calcMod.optimize()

In [50]:
calcMod.objVal

650.0

In [51]:
calcModX.x 

array([100., 170.])

**Answer:**
The maximum profit the company can gain is $650, by selling 100 scientific and 170 graphing calculators per day. 